retention/cohort_analysis.py

```diff
@@ -528,26 +588,39 @@ def generate_by_bucket(
                                                                                         cohort_end_date)
 
         bucket_num = bucket + 1
-        client_types = get_cohort_clients_bucket(cohort_users, users_and_devices_to_client,
-                                                 bucket_start_date, bucket_end_date)
-        cohorts.append((ts_to_str(cohort_start_date), bucket_num, client_types))
-
-    return cohorts
+        client_types = get_cohort_clients_bucket(
+            cohort_users,
+            cohort_start_date,
+            users_and_devices_to_client,
+            bucket_start_date,
+            bucket_end_date,
+        )
+        for cohort_key, count in client_types:
+            yield cohort_key, bucket_num, count
 
 
-def write_to_mysql(table, buckets_stats, dry_run):
+def write_to_mysql(table: str, buckets_stats: CohortStatsResult, dry_run: bool):
     statements_and_values = []
-    for cohort_date, bucket_num, client_counts in buckets_stats:
-        for client in [ELEMENT_ANDROID, RIOTX_ANDROID, ELEMENT_ELECTRON, ELEMENT_IOS, WEB]:
-            insert_bucket = f"INSERT INTO {table} (date, client, b{bucket_num}) VALUES" \
-                            f"(%s, %s, %s) " \
-                            f"ON DUPLICATE KEY UPDATE b{bucket_num}=VALUES(b{bucket_num});"
-            statements_and_values.append((insert_bucket, cohort_date, client, client_counts[client]))
+
+    for cohort_key, bucket_num, count in buckets_stats:
+        insert_bucket = f"""\
+            INSERT INTO {table} (date, client, sso_idp, b{bucket_num})
+            VALUES (%s, %s, %s, %s)
+            ON DUPLICATE KEY UPDATE b{bucket_num}=VALUES(b{bucket_num})
+        """
+        statements_and_values.append((
+            insert_bucket, (
+                cohort_key.cohort_start_date,
+                cohort_key.client_type,
+                cohort_key.sso_idp,
+                count
+            )
+        ))
 
     if dry_run:
         logging.info("Would have run the following SQL statements")
-        for insert_string, cohort_date, client, count in statements_and_values:
-            print(insert_string % (cohort_date, client, count))
+        for insert_string, values in statements_and_values:
+            print(insert_string % values)
```

Comment:
Suggest either removing the print or using the logger.

Reply:
a dry run would be a bit useless if we didn't print at all here.

It seems like this is the primary output of the script in dry-run mode, so you'd expect it to be written out directly to `stdout` rather than being timestamped by the logger and written to `stderr`, so I think that `print` is correct.